src/kernels/ScalarAdvectionArtDiff.C

```diff
@@ -52,11 +58,39 @@ ScalarAdvectionArtDiff::ScalarAdvectionArtDiff(const InputParameters & parameter
 }
 
 Real
-ScalarAdvectionArtDiff::computeQpResidual()
+ScalarAdvectionArtDiff::tau()
 {
   RealVectorValue U(_u_vel[_qp], _v_vel[_qp], _w_vel[_qp]);
+  Real u_norm = U.norm();
+  Real h;
+  if (_mesh.dimension() == 1)
+    h = _current_elem->volume();
+  else if (_mesh.dimension() == 2)
+    h = std::sqrt(_current_elem->volume());
+  else if (_mesh.dimension() == 3)
+    h = std::cbrt(_current_elem->volume());
+  Real gamma = u_norm * h / 2. / _D[_qp];
+  if (gamma <= 1e-10)
+    return 0.; // low-Peclet flow => no artificial diffusion
+  else if (gamma >= 5.)
+    return 1. - 1. / gamma; // prevent overflow
+
+  return 1. / std::tanh(gamma) - 1. / gamma;
+}
 
-  Real delta = U.norm() * _current_elem->hmax() / 2. * _tau;
+Real
+ScalarAdvectionArtDiff::computeQpResidual()
+{
+  RealVectorValue U(_u_vel[_qp], _v_vel[_qp], _w_vel[_qp]);
+  Real volume = _current_elem->volume();
+  Real h;
+  if (_mesh.dimension() == 1)
+    h = _current_elem->volume();
+  else if (_mesh.dimension() == 2)
+    h = std::sqrt(_current_elem->volume());
+  else if (_mesh.dimension() == 3)
+    h = std::cbrt(_current_elem->volume());
+  Real delta = U.norm() * h / 2. * ScalarAdvectionArtDiff::tau();
```

Comment:
Thanks for reviewing my PR!

I changed it to this ```h``` because the derivation calls for the characteristic length. For regular quadratic/cubic elements with equal sides, ```h``` is identical to the largest side of the element. For skewed elements, the largest side of the element is a bad estimator for the characteristic length. Granted, the derivation does not explicitly cover skewed elements in 2D/3D, but using the characteristic length would be more accurate for this sort of isotropic diffusion effect which does not take into account the velocity direction.